compiler/closureiters.nim

```diff
@@ -384,7 +384,8 @@ proc hasYieldsInExpressions(n: PNode): bool =
   of nkSkip:
     discard
   of nkStmtListExpr:
-    if isEmptyType(n.typ):
+    if isEmptyType(n.typ) or (n[^1].kind in nkCallKinds and
+          n[^1][0].kind == nkSym and sfDiscardable in n[^1][0].sym.flags):
```

Comment:
But you don't know that just because it's discardable it also actually was discarded! This needs to be changed in the sem'checker instead.

Reply:
closureiterators or semcheck, which is the first phase?

Reply:
semcheck is the first phase. And yeah, it needs to change the n.typ to `nil` (preferable) or else rewrites the AST to contain a ``discard``  statement.

Reply:
sem phase

```json
{
  "kind": "nkDiscardStmt",
  "typ": "nil",
  "sons": [{
    "kind": "nkStmtListExpr",
    "flags": ["nfSem"],
    "typ": "tyString",
    "sons": [{
      "kind": "nkYieldStmt",
      "flags": ["nfSem"],
      "typ": "nil",
      "sons": [{
        "kind": "nkIntLit",
        "flags": ["nfSem"],
        "typ": "tyInt",
        "intVal": 12
      }]
    }, {
      "kind": "nkCall",
      "flags": ["nfSem"],
      "typ": "tyString",
      "sons": [{
        "kind": "nkSym",
        "typ": "tyProc",
        "sym": {
          "kind": "skProc",
          "name": "hello",
          "id": 369098753,
          "flags": ["sfUsed", "sfGlobal", "sfDiscardable"]
        }
      }]
    }]
  }]
}
```

Reply:
It is already done
```json
{
  "kind": "nkDiscardStmt",
  "typ": "nil",
  "sons": [{
    "kind": "nkStmtListExpr",
    "flags": ["nfSem"],
    "typ": "tyString",
    "sons": [{
      "kind": "nkYieldStmt",
      "flags": ["nfSem"],
      "typ": "nil",
      "sons": [{
        "kind": "nkIntLit",
        "flags": ["nfSem"],
        "typ": "tyInt",
        "intVal": 12
      }]
    }, {
      "kind": "nkCall",
      "flags": ["nfSem"],
      "typ": "tyString",
      "sons": [{
        "kind": "nkSym",
        "typ": "tyProc",
        "sym": {
          "kind": "skProc",
          "name": "hello",
          "id": 369098753,
          "flags": ["sfUsed", "sfGlobal", "sfDiscardable"]
        }
      }]
    }]
  }]
}
discard
  yield 12
  hello()
```

Reply:
I see